### DISCUSSION TITLE
Leniency for generic types with [StructLayout(LayoutKind.Explicit)]

### DISCUSSION THREAD

**Discussion by tagcode** [dotnet/runtime#108063] [General]
**[StructLayout(LayoutKind.Explicit)]** isn't supported for generic types. I think a little bit leniency could be added here, if the field with generic type **T** is the last field, OR if there is no **T** field at all and generics is for properties and methods only.

```C#
using System.Runtime.InteropServices;

MyStruct<long> myStruct = new MyStruct<long>();
// ^ thrown at run-time: System.TypeLoadException: 'generic types cannot have explicit layout.'

[StructLayout(LayoutKind.Explicit)]
public struct MyStruct<T> where T : struct
{
    [FieldOffset(0)]
    public long Type;
    [FieldOffset(0)]
    public ushort Id;
    [FieldOffset(8)]
    public T Value;
}
```

**Comment by behdad088** [comment]


**Reply by tagcode** [reply]
My actual case was this:

```c#
[StructLayout(LayoutKind.Explicit)]
public struct MyStruct<T> where T : struct
{
    /// <summary>
    /// 0 for <see cref="ReadOnlyMemory{T}"/>
    /// 1 for <see cref="Memory{T}"/>
    /// </summary>
    [FieldOffset(0)]
    public long Type;

    [FieldOffset(8)]
    public Memory<T> Memory;
    [FieldOffset(8)]
    public ReadOnlyMemory<T> ReadOnlyMemory;
}
```

In current workaround I have duplicate Memory fields.
```C#
public struct CurrentWorkaround<T>
{
    public Memory<T> Memory;
    public ReadOnlyMemory<T> ReadOnlyMemory;
}
```

1. I can't have every possible T
2. Ok, I tried your suggestion (below).
3. Nah

```C#
Suggestion2<char> a = new Suggestion2<char> { AsReadOnlyMemory = "Hello world".AsMemory() };
Suggestion2<char> b = new Suggestion2<char> { AsMemory = "Hello world".ToArray() };

Console.WriteLine(a.AsReadOnlyMemory);
Console.WriteLine(b.AsMemory);

public struct Suggestion2<T>
{
    public ReadOnlyMemory<T> AsReadOnlyMemory
    {
        get => Type == 0 ? memory : (ReadOnlyMemory<T>)AsMemory;
        set { memory = value; Type = 0; }
    }

    public Memory<T> AsMemory
    {
        get => Type == 1 ? Unsafe.As<ReadOnlyMemory<T>, Memory<T>>(ref memory) : throw new InvalidOperationException($"Not assiged as {typeof(Memory<T>)}.");
        set {
            memory = Unsafe.As<Memory<T>, ReadOnlyMemory<T>>(ref value);
            Type = 1;
        }
    }

    ReadOnlyMemory<T> memory;

    /// <summary>
    /// 0 for <see cref="ReadOnlyMemory{T}"/>
    /// 1 for <see cref="Memory{T}"/>
    /// </summary>
    public byte Type;
}
```

This works with slightly-iffy assumption that Memory<T> and ReadOnlyMemory<T> always have same layout so that GC doesn't get messed up with their ._object fields as they both have it at same location. GC should release Memory<T>._object from ReadOnlyMemory<T>._object struct just the same.

**Reply by huoyaoyuan** [reply]
> Memory and ReadOnlyMemory always have same layout so that GC doesn't get messed up with the object field

Although this is correct, it may not be easy to reveal in the order of type loading.

Generally, using explicit layout to cast between types is not recommended and may cause issues like false address exposing. `Unsafe.BitCast` and `Unsafe.As` are created for these type of usage.

**Reply by tagcode** [reply]
> Generally, using explicit layout to cast between types is not recommended and may cause issues like false address exposing. `Unsafe.BitCast` and `Unsafe.As` are created for these type of usage.

I wasn't casting with layout, simply using it as a one-of type switch.

**Reply by huoyaoyuan** [reply]
A union-like type switch isn't friendly for the JIT and GC. Always storing in one concrete type and cast for accessing is the correct approach.

**Reply by tagcode** [reply]
> A union-like type switch isn't friendly for the JIT and GC. Always storing in one concrete type and cast for accessing is the correct approach.

I know it is iffy. I'm not going with layout on the the actual case I had. It is not relevant to the original posting.

**Reply by tagcode** [reply]
> it's not possible due to the dynamic nature of generics and the need to know the exact memory layout, but they are some workout that might be a bit useful for you:

I was thinking that at runtime it would be possible, because the exact layout is known at loading time because all the type arguments are known.

It could be possible in some cases, for example if there is only one field whoose type uses generic type parameter, and no other fields after this.

**Reply by huoyaoyuan** [reply]
> I was thinking that at runtime it would be possible, because the exact layout is known at loading time because all the type arguments are known.

Yes it's known, but it does require a non-trivial algorithm to prove it's safe, and correctly report it to GC. The use case doesn't worth the change.

**Reply by jnm2** [reply]
What if it is stored as Memory<T> or ReadOnlyMemory<T> always, and then if you ever need to go in the opposite direction, you can use [`MemoryMarshal.AsMemory(readOnlyMemory)`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.memorymarshal.asmemory)?

**Reply by tagcode** [reply]
> What if it is stored as Memory or ReadOnlyMemory always, and then if you ever need to go in the opposite direction, you can use [`MemoryMarshal.AsMemory(readOnlyMemory)`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.memorymarshal.asmemory)?

Now this solves it. Since it is in standard library, it is safely maintained. 

Albeit, it seems to be implemented with pointer-casting, the exact same way I was casting myself in one suggestion.

```C#
	public static Memory<T> AsMemory<T>(ReadOnlyMemory<T> memory)
	{
		return Unsafe.As<ReadOnlyMemory<T>, Memory<T>>(ref memory);
	}
```


But atleast this way MS takes responsibility for future compability.

**Reply by tagcode** [reply]
> This works with slightly-iffy assumption that Memory and ReadOnlyMemory always have same layout so that GC doesn't get messed up with their ._object fields as they both have it at same location. GC should release Memory._object from ReadOnlyMemory._object struct just the same.

So it is not iffy after all since [MemoryMarshal.AsMemory](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.memorymarshal.asmemory) is implemented with the same assumption.

**Reply by colejohnson66** [reply]
Is this reply AI-generated? Because `unsafe` (the keyword) code provides no help here.

**Reply by tagcode** [reply]
> Is this reply AI-generated? Because `unsafe` (the keyword) code provides no help here.

Pointer based type-casting can be done within unsafe block just the same as with the Unsafe class and MemoryMarshal.

```C#
Suggestion2B<char> a = new Suggestion2B<char> { AsReadOnlyMemory = "Hello world".AsMemory() };
Suggestion2B<char> b = new Suggestion2B<char> { AsMemory = "Hello world".ToArray() };

Console.WriteLine(a.AsReadOnlyMemory);
Console.WriteLine(b.AsMemory);

public struct Suggestion2B<T>
{
    public ReadOnlyMemory<T> AsReadOnlyMemory
    {
        get => Type == 0 ? memory : (ReadOnlyMemory<T>)AsMemory;
        set { memory = value; Type = 0; }
    }

    public unsafe Memory<T> AsMemory
    {
        get {
            if (Type != 1) throw new InvalidOperationException($"Not assiged as {typeof(Memory<T>)}.");
            fixed (void* ptr = &memory) return *(Memory<T>*)ptr;
        }
        set {
            memory = *(ReadOnlyMemory<T>*)&value;
            Type = 1;
        }
    }

    ReadOnlyMemory<T> memory;

    /// <summary>
    /// 0 for <see cref="ReadOnlyMemory{T}"/>
    /// 1 for <see cref="Memory{T}"/>
    /// </summary>
    public byte Type;
}
```

**Reply by tagcode** [reply]
> Is this reply AI-generated? Because `unsafe` (the keyword) code provides no help here.

Now that I read it second time, the layout of the text looks very AI-like.

**Reply by tagcode** [reply]
Edited: On side note.

If generics was possible, this is what a layout type casting would look like. Provided as exploration to type-casting options.

```C#
[StructLayout(LayoutKind.Explicit)]
public struct Switcharoo<T>
{
    [FieldOffset(0)]
    public Memory<T> Memory;
    [FieldOffset(0)]
    public ReadOnlyMemory<T> ReadOnlyMemory;
}
```

It works just fine without generics
```C#
var x = Switcharoo { Memory = "Hello World".ToArray() };
ReadOnlyMemory<char> rom = x.ReadOnlyMemory;

[StructLayout(LayoutKind.Explicit)]
public struct Switcharoo
{
    [FieldOffset(0)]
    public Memory<char> Memory;
    [FieldOffset(0)]
    public ReadOnlyMemory<char> ReadOnlyMemory;
}
```

Type safety can be added by making the fields private and adding the same properties as earlier examples.

**Reply by jnm2** [reply]
Let me throw just one more thing in here: Why even have a ReadOnlyMemory member? `Memory<T>` implicitly converts to `ReadOnlyMemory<T>`.

**Reply by tagcode** [reply]
> Let me throw just one more thing in here: Why even have a ReadOnlyMemory member? `Memory<T>` implicitly converts to `ReadOnlyMemory<T>`.

For safety, I don't want to be able to accidentally read an immutable string as Memory<char>.

For storing as field, all solutions are the same because Memories are pointer castable.

If you were referring to the Switcharoo example, the ReadOnlyMemory -> Memory casting is the one without implicit operator. With that example I was referring to type-casting options, and it was interesting to try out if it worked.

**Reply by jnm2** [reply]
> I don't want to be able to accidentally read an immutable string as Memory.

But since there is still a `Memory<T>` member that does not throw if the `Type` is incorrect, what is keeping you from accidentally reading an immutable string as `Memory<char>`?

**Reply by tagcode** [reply]
> > I don't want to be able to accidentally read an immutable string as Memory.
> 
> But since there is still a `Memory<T>` member that does not throw if the `Type` is incorrect, what is keeping you from accidentally reading an immutable string as `Memory<char>`?

Ugh, I edited the text lines in the Switcharoo.

**Reply by elgonzo** [reply]
> So it is not iffy after all since [MemoryMarshal.AsMemory](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.memorymarshal.asmemory) is implemented with the same assumption.

No, for the CoreCLR team it is obviously not an assumption. Because the CoreCLR team controls/defines both the the memory layout of Memory\<T\> and ReadOnlyMemory\<T\> as well as the implementation of MemoryMarshal.AsMemory, and they can change anything regarding these in any way they see fit for any new .NET release.

You don't control either of these, and such changes wouldn't be publicly telegraphed in an obvious manner (you basically would need to continually monitor CoreCLR development to maintain awareness of such changes). So you _assume_ that the current status quo regarding the current memory layout and the current implementation of MemoryMarshal.AsMemory won't change. Depending on how long into the future you intend to maintain your project, your initial judgement of it being a slightly-iffy assumption appears quite reasonable.

**Reply by tagcode** [reply]
> > So it is not iffy after all since [MemoryMarshal.AsMemory](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.memorymarshal.asmemory) is implemented with the same assumption.
> 
> No, for the CoreCLR team it is obviously not an assumption. Because the CoreCLR team controls/defines both the the memory layout of Memory<T> and ReadOnlyMemory<T> as well as the implementation of MemoryMarshal.AsMemory, and they can change anything regarding these in any way they see fit for any new .NET release.
> 
> You don't control either of these, and such changes wouldn't be publicly telegraphed in an obvious manner (you basically would need to continually monitor CoreCLR development to maintain awareness of such changes). 

Within the context of already published standard libraries, the layout are not just "assumptions" but written facts. 

> So you _assume_ that the current status quo regarding the current memory layout and the current implementation of MemoryMarshal.AsMemory won't change. Depending on how long into the future you intend to maintain your project, your initial judgement of it being a slightly-iffy assumption appears quite reasonable.

As per future compability I had already stated: "But atleast this way MS takes responsibility for future compability [of using MemoryMarshal]." expressing that I was aware that future API compability can only go though API call. 

After which I though it would be so plainly obvious that it would not needed to be vented any more, but you started _assuming_ what I assumed.

And in practical terms, I seriously doubt that CoreCLR team actually can divert the layout of those two types even if they wanted, because something somewhere will break. Tiniest changes have tendency to propagate all over the place.

> > So it is not iffy after all since [MemoryMarshal.AsMemory](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.memorymarshal.asmemory) is implemented with the same assumption.

But yes, the last word here is poorly phrased. I've been recovering from three days of fewer and I was making typos and having wrong word choices all over the place. Every word must be exact or else somebody flips their mind.

**Reply by tagcode** [reply]
Conversation gets "white noise" from here on. This doesn't serve any useful purpose, only personal need to bicker about wording and irrelevant points.

**Reply by elgonzo** [reply]
> Conversation gets "white noise" from here on. This doesn't serve any useful purpose, only personal need to bicker about wording and irrelevant points.

Someone invent a mirror, please...

**Reply by tagcode** [reply]
> > Conversation gets "white noise" from here on. This doesn't serve any useful purpose, only personal need to bicker about wording and irrelevant points.
> 
> Someone invent a mirror, please...

You are there too pal. The point was made for other readers.

**Reply by jnm2** [reply]
@tagcode I can't think of a way in which it is more safe to have a `Memory<T>` union member than to have only a `ReadOnlyMemory<T>` member and to use MemoryMarshal on it. You can access both things without regard to the struct's Type field.

**Comment by jnm2** [comment]
For the provided example, here's something you can do today:
```cs
public struct MyStruct<T> where T : struct
{
    public MyStructHeader Header;
    public T Value;

    [StructLayout(LayoutKind.Explicit)]
    public struct MyStructHeader
    {
        [FieldOffset(0)]
        public long Type;
        [FieldOffset(0)]
        public ushort Id;
    }
}
```
I see at your https://github.com/dotnet/runtime/discussions/108063#discussioncomment-10706663 comment that the union in your actual use case is between two `T`-dependent fields, so this wouldn't work for that.

**Comment by hez2010** [comment]
@jkotas @AaronRobinsonMSFT I think we can remove this restriction now? We already have code to validate if the type is blittable, and it will throw another exception during type loading if it's not, so there should be no issue around GC reporting. With generic support in function pointers now I can see some use cases on this. 
This can also help us remove some workaround in the BCL like `EntryInfo` in `GenericCache.cs`.

I can make a PR for this. While DO note that this restriction comes from `CLI - Partition 1, section 9.5 - Generic types shall not be marked explicitlayout`, by allowing this we will need to alter the spec as well.

**Reply by jkotas** [reply]
I do not see a problem with requiring a helper type for overlapping non-generic fields in generic types. I think the current restriction where we are disallowing explicit layout for generics is fine and easy to explain.

**Reply by colejohnson66** [reply]
If it is to be kept as disallowed, I would argue that the compiler should error (or at least warn) on these situations.

**Reply by AaronRobinsonMSFT** [reply]
> I think the current restriction where we are disallowing explicit layout for generics is fine and easy to explain.

Agree. Changing how this works doesn't seem to offer all that much and creates places where more details about interop are necessary. If this isn't for an interop scenario, then I would argue we should be improving the JIT and type loader so uses of `LayoutKind.Explicit` aren't necessary.

> If it is to be kept as disallowed, I would argue that the compiler should error (or at least warn) on these situations.

@colejohnson66 I'm going to assume you would be okay with this being a new analyzer?

**Reply by colejohnson66** [reply]
@AaronRobinsonMSFT Yep. An analyzer would probably be fine, but I'd also argue it should be enabled by default. Anything that compiles, but *unconditionally* throws at runtime should either be illegal to write, or warned about.

**Reply by jkotas** [reply]
> Anything that compiles, but unconditionally throws at runtime 

There are *many* constructs like that. It does not make sense to duplicate all error handling from the runtime and libraries in analyzers so that you get a warning at compile time for these constructs.

Issuing warning at compile time for a construct that unconditionally throw at runtime only makes sense if it is very frequent one.
